File: experiments/runs/2026-09-14_18-05-04_claim-office-example-mapping_external-tcrdd-bsene-2026-09-14-cc_opus-5-requesty-no-thinking/src/policy.ts
import type { Item } from './quote';

const INSURANCE_VALUE: Record<string, number> = {
  sword: 1000,
  amulet: 600,
  staff: 800,
  potion: 400,
  rune: 250,
  moonstone: 250,
};

const DEDUCTIBLE = 100;
const CAP_MULTIPLIER = 2;
const HIGH_ENCHANTMENT_LEVEL = 8;
const HIGH_ENCHANTMENT_REIMBURSEMENT = 0.5;

export interface Damage {
  itemType: string;
  amount: number;
}

export interface ClaimResult {
  payout: number;
  remainingCap: number;
}

export function insuranceSum(items: Item[]): number {
  return items.reduce((sum, item) => {
    const value = INSURANCE_VALUE[item.type];
    if (value === undefined) {
      throw new Error(`unknown item type: ${item.type}`);
    }
    return sum + value;
  }, 0);
}

/**
 * Damage to a highly enchanted item is reimbursed at half; dragon
 * material is reimbursed in full, which is already the default, so the
 * enchantment clause wins whenever both would apply.
 */
function reimbursableAmount(item: Item, damage: number): number {
  return (item.enchantment ?? 0) >= HIGH_ENCHANTMENT_LEVEL
    ? damage * HIGH_ENCHANTMENT_REIMBURSEMENT
    : damage;
}

export class Policy {
  private cap: number;

  constructor(private readonly items: Item[]) {
    this.cap = CAP_MULTIPLIER * insuranceSum(items);
  }

  /**
   * Settles one damage event. A damage entry that names an item the
   * policy does not cover — or that exceeds the number of copies
   * insured — rejects the whole claim, leaving the cap untouched.
   */
  claim(damages: Damage[]): ClaimResult {
    const uncovered = [...this.items];
    let payout = 0;

    for (const damage of damages) {
      if (damage.amount < 0) {
        throw new Error(`negative damage amount: ${damage.amount}`);
      }
      const index = uncovered.findIndex((item) => item.type === damage.itemType);
      if (index === -1) {
        throw new Error(`item not covered by the policy: ${damage.itemType}`);
      }
      const [item] = uncovered.splice(index, 1);
      // The deductible applies once per damaged item.
      payout += reimbursableAmount(item, damage.amount) - DEDUCTIBLE;
    }

    // Rounded down, in the MHPCO's favour, then limited by the cap.
    payout = Math.min(Math.floor(payout), this.cap);
    this.cap -= payout;
    return { payout, remainingCap: this.cap };
  }
}

File: experiments/runs/2026-09-20_07-28-04_claim-office-java-example-mapping_exact-ptdd-v1-cc_opus-5-no-thinking/src/main/java/Policy.java
import java.math.BigDecimal;
import java.util.List;
import java.util.Map;

/** The items one MHPCO policy covers, and the payout cap remaining on it. */
class Policy {

    private static final BigDecimal COMPONENT_INSURANCE_VALUE = new BigDecimal("250");
    private static final BigDecimal CAP_MULTIPLE_OF_INSURANCE_SUM = new BigDecimal("2");

    private static final Map<String, BigDecimal> INSURANCE_VALUES = Map.of(
            "sword", new BigDecimal("1000"),
            "amulet", new BigDecimal("600"),
            "staff", new BigDecimal("800"),
            "potion", new BigDecimal("400"),
            "rune", COMPONENT_INSURANCE_VALUE,
            "moonstone", COMPONENT_INSURANCE_VALUE);

    private final List<Map<String, Object>> insuredItems;
    private BigDecimal remainingCap;

    Policy(List<Map<String, Object>> insuredItems) {
        this.insuredItems = List.copyOf(insuredItems);
        this.remainingCap = insuranceSum().multiply(CAP_MULTIPLE_OF_INSURANCE_SUM);
    }

    /** Every reported damage must refer to an item this policy actually covers. */
    void verifyCovers(List<Map<String, Object>> damages) {
        for (Map<String, Object> damage : damages) {
            String itemType = (String) damage.get("itemType");
            if (damagesOfType(damages, itemType) > countInsuredOfType(itemType)) {
                throw new IllegalArgumentException(
                        "policy does not cover that many items of type " + itemType);
            }
        }
    }

    private long damagesOfType(List<Map<String, Object>> damages, String itemType) {
        return damages.stream().filter(damage -> itemType.equals(damage.get("itemType"))).count();
    }

    private long countInsuredOfType(String itemType) {
        return insuredItems.stream().filter(item -> itemType.equals(item.get("type"))).count();
    }

    Map<String, Object> insuredItemOfType(String itemType) {
        for (Map<String, Object> item : insuredItems) {
            if (itemType.equals(item.get("type"))) {
                return item;
            }
        }
        return Map.of();
    }

    /** The total payout per policy is capped at twice the insurance sum. */
    BigDecimal limitToRemainingCap(BigDecimal desiredPayout) {
        return desiredPayout.min(remainingCap);
    }

    void recordPayout(BigDecimal payout) {
        remainingCap = remainingCap.subtract(payout);
    }

    BigDecimal remainingCap() {
        return remainingCap;
    }

    private BigDecimal insuranceSum() {
        BigDecimal sum = BigDecimal.ZERO;
        for (Map<String, Object> item : insuredItems) {
            sum = sum.add(INSURANCE_VALUES.get(item.get("type")));
        }
        return sum;
    }
}
